Sorter database integration

redshiftzero opened the pull request:
Implements #26. Adds options in `config.ini` to upload the onions into the database when the sorter runs. One can also turn the `upload_to_db` option to false to run the sorter as before without uploading the data into the database (which one should do if one is testing things). Either way, the onions will also be saved in `logging` in a pickle file. 

This PR also adds SQL scripts in `fpsd/db` to setup the database. There is a graphviz diagram showing the database design in `doc/images`.

redshiftzero committed:
Add database design graphviz diagram

redshiftzero committed:
Add libpq-dev to packages that ansible installs

redshiftzero committed:
Add new options to config.ini

redshiftzero committed:
Add database upload code

redshiftzero committed:
Update pip requirements

conorsch commented on fpsd/db/create_db.sql:
```diff
@@ -0,0 +1,2 @@
+CREATE USER fp_user WITH PASSWORD 'thisisnottherealpasswordcomeonyall';
```
:joy:

redshiftzero committed:
Add function for processing datetime objects from new config options

redshiftzero committed:
Add tests for new utils

redshiftzero committed:
Add methods to RawStorage to handle upload of traces to database

redshiftzero committed:
Enable crawler to save data to the database

redshiftzero commented:
These new commits implement issue #25, also integrating the crawler with the database. 

These are new config options for the crawler:
• `use_database`: select crawler onions from the database instead of using the pickle file, and insert traces back into the database. if False, they will be saved as before in text files
• `hs_history_lookback`: define the timespan from which to pick the onions to crawl (e.g. those HS seen up as of the last hour, last day, last week?). Some helper functions and corresponding tests are added to `utils.py` and `test/test_utils.py` to handle this. 

What happens in `crawler.py` when `use_database` is set to True:
• In `Crawler.__init__`, the crawler will insert a row into `raw.crawls` storing some general information about this crawl, such as the os and entry node that is used. The database insert is done in `RawStorage.add_crawl()` in `database.py`. 
• In `Crawler.collect_onion_trace()`, the crawler will insert a row corresponding to a given trace into `frontpage_examples` and will then bulk insert each measured cell into `frontpage_traces`. 
• Note that `class_data` will still be an `OrderedDict` (to maintain the functionality of the pickle file). However, it will now contain two dicts for the non-monitored and monitored classes. For comparison, when the crawler uses a pickle file, the form of `class_data` is an `OrderedDict` with two sets for the non-monitored and monitored classes. The logic in `crawl_monitored_nonmonitored_classes` is changed slightly to handle these two cases.

fowlslegs commented:
Can you rebase this on master? Also, I see you added `psycopg2` to `sorter-requirements.in`, but don't import it anywhere?

redshiftzero commented:
ah, `psycopg2` is the Python DBAPI for PostgreSQL that SQLalchemy needs to work with the database.

fowlslegs commented:
Okay, so it's listed as an optional and not hard requirement of the SQLAlchemy package, and while we don't use it directly, we can't communicate w/ the Postgres DB unless we have it? Am I getting that right?

redshiftzero commented:
yep, that's correct

redshiftzero commented:
Rebased on master for review

Commit:
Add database role, fpsd db, raw schema & assoc tables

- Installs and starts PostgresSQL server & initializes the database cluster.
- Creates the fp_user role, fpsd database, and the raw schema.
- Creates the crawls, hs_history, frontpage_examples, and frontpage_traces
  tables for the raw schema.
- Creates a random password for the fp_user and saves it to a PGPASSFILE.
  - Documents the creation of this file in the README.

fowlslegs commented:
I just automated the setup of the database for testing. Can you try to run it for me and then re-provision as well @redshiftzero:

```
vagrant destroy -f
vagrant up
vagrant provision
```

fowlslegs commented:
Be sure to pull first and also follow the instructions I just added to the README on setting up a virtualenv from which to run the vagrant commands, as you need specific Python and Ansible versions for everything to work.

redshiftzero commented:
I followed these instructions and re-provisioned: the database seems to be getting set up with no problems in the VM 🎆

redshiftzero commented on README.md:
```diff
@@ -123,6 +131,13 @@ A systemd unit is also provided to run the crawler on repeat. Simply run
+
+A random password will be generated for you automatically and will be saved to
+`/tmp/passwordfile` on the Ansible controller (your host machine), and saved to
+a `PGPASSFILE`, `/var/lib/postgres/pgpass.conf`, on the remote host (the VM
```
it looks like the password is stored in `/var/lib/postgresql/pgpass.conf`

psivesely replied:
yah

fowlslegs commented:
Notes on individual commits:
* https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/3d94ceb7ff51d5da65b7ef7be0ba614215a44064 should be its own PR.
* https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/a1324ec743ae6624631af701f78dd84d3e7e096b should either be removed or they should be deleted and then melded into https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/d3407320d8c3d09f8741f0189073049309de6932 using fixup during a rebase.
* Not sure if https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/a7d6ae2c479a5d3c01a759bb1fdbbe77b0eb42aa is needed. Maybe we can just add some info on setting env vars in the README or just Ansible-ize it?
* I just barely modified https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/a7dbe9b657d52503cc271cabbf9c8c899b7a5432 (now 3ee61fa22e9b6c4fd8d30969a977ba917fc98db5) in a rebase.
* Something unrelatedish snuck into https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/c17c617ef570923c048ec4ee2f4e03cf33d36382. I think it's okay to leave in, but would be neater to be it's own commit or melded into https://github.com/freedomofpress/FingerprintSecureDrop/pull/27/commits/a1324ec743ae6624631af701f78dd84d3e7e096b.

Some other notes:
* We should write more for the README on the database and include that dank Graphviz graph you made.

fowlslegs commented:
I meant to figure out a solution to this before pushing, but in my last commit here, I realized that if you try to re-provision a VM after reboot `/tmp/passwordfile` will be gone, so it will be re-generated and then Ansible will generate a new password which will get saved to the `/var/lib/postgresql/pgpass.conf`. That password won't work and will have erased the working one. Rather than worry about storing it permanently on the controller, which would become too complicated given the possibility of deploying to multiple VPSs, we should use a `when` statement to not re-generate a password when `pgpass.conf` is present.

Another thing to figure out is how to support provisioning a VM that doesn't create it's own local SQL server, but instead is configured to write to a remote database.

Commit:
Ensure idempotency of db initialization tasks

If /etc/passwordfile is no longer on the controller this makes sure:
- fp_user is not re-created with new password.
- The password file on the remote host is not modified.

redshiftzero committed:
Update README

redshiftzero committed:
Update sorter to enable saving into the database

redshiftzero commented:
Thanks for the feedback - responding bullet-by-bullet:

> 3d94ceb should be its own PR.

I see that you merged in 3d94ceb separately so we are good there.

> a1324ec should either be removed or they should be deleted and then melded into d340732 using fixup during a rebase.

I’m assuming that the database files in a1324ec should be removed because they’ve now been committed under `roles/database/files`. I’ve removed the entire `db` directory since this is being done by Ansible. I can add back creating the `features` schema at a later time. 

> Not sure if a7d6ae2 is needed. Maybe we can just add some info on setting env vars in the README or just Ansible-ize it?

I see your point here, I’ve removed this file and rewritten `RawStorage().__init__()` to directly use the environmental variables. I added a note to README.md that these environmental variables must be set in order to use the database. 

> I just barely modified a7dbe9b (now 3ee61fa) in a rebase.

Looks good.

> Something unrelatedish snuck into c17c617. I think it's okay to leave in, but would be neater to be it's own commit or melded into a1324ec.

Fixed

> We should write more for the README on the database and include that dank Graphviz graph you made 

I wrote a little more on the README describing the database and put in the graphviz figure. I hope it is “dank” 🏄

psivesely commented on roles/database/tasks/initialize_raw_schema.yml:
```diff
@@ -0,0 +1,29 @@
+  command: psql -c '{{ lookup("file", item) }}'
+  environment: "{{ psql_env }}"
+  with_items: "{{ raw_schema_tables }}"
+  when: "item[13:-4] not in tables.stdout"
```
A comment here would help understanding.

psivesely commented on README.md:
```diff
@@ -74,10 +74,18 @@ service.
 This repo uses submodules, so use `git clone --recursive`. You will also need to
 install the following:
```
Maybe something like "this was successfully tested on..." instead of a hard you must have these versions. Ansible must be >2 fasho.... Also Python 2.7 must be your default system version or you must use a virtualenv.

psivesely commented on README.md:
````diff
@@ -123,6 +131,31 @@ A systemd unit is also provided to run the crawler on repeat. Simply run
+
+The data collection programs - the sorter and crawler - are integrated with a PostgreSQL database. When the `use_database` option is set to True in the sorter part of `config.ini`, the sorter will save its sorted onion addresses in the database. When the `use_database` option is set to True in the crawler part of `config.ini`, the crawler will grab onions from the database, connect to them, measu ...
+
+```
````
The user should not have to set these. Instead we can export them in `/etc/profile` or `/etc/profile.d/`. Also, `PGPASSWORD` should be avoided in favor of `PGPASSFILE`.

Commit:
Use standard naming scheme for initialize-raw-schema tasklist

Commit:
Set psql env vars in /etc/profile.d/fpsd-database.sh

Commit:
Add info string to PGPASSFILE template

Commit:
Specify fpsd database when querying schemas

fowlslegs commented:
So `/etc/profile` and `/etc/profile.d/*` are only sourced by login shells. So you need to do `sudo -Es && su postgres` to keep the env vars if you need to operate as the postgres user with the way things are set up currently. Since we shouldn't need to support other shells, it's probably better that these are set in `/etc/bash.bashrc` so that they're always accessible to all bash users. ~~Since there is already stuff in the there though, and we want to use a template for the file, it might make sense to leave the `/etc/profile.d/fpsd-database.sh` script where it is and then ensure the line `source /etc/profile.d/fpsd-database.sh` is present in `/etc/bash.bashrc.~~ Nevermind, I just thought of a way you can do this just using `lineinfile`.

Commit:
Use /etc/bash.bashrc to set env vars instead of /etc/profile.d/...

So /etc/profile and /etc/profile.d/* are only sourced by login shells. So you
need to do sudo -Es && su postgres to keep the env vars if you need to operate
as the postgres user with the way things are set up currently. Since we
shouldn't need to support other shells, it's probably better that these are set
in /etc/bash.bashrc so that they're always accessible to all bash users.

conorsch commented on roles/database/tasks/initialize-fpsd-db.yml:
```diff
@@ -43,11 +42,12 @@
     mode: "0644"
     owner: root
     group: root
+  with_items: "{{ psql_env.items() }}"
```
Clever, but doesn't work as expected. Creates the following block in `/etc/bash.bashrc`:

```
export P=G
export f=p
export /=v
export l=o
export 5=4
```

Not what we want. Ansible has a built-in loop parameter with working with dict vars. Try this:

```
- name: Set the database environment variables in `/etc/bash.bashrc`.
  become: true
  lineinfile:
    line: 'export {{ item.key }}={{ item.value }}'
    regexp: '^export {{ item.key }}='
    dest: /etc/bash.bashrc
    mode: "0644"
    owner: root
    group: root
  with_dict: "{{ psql_env }}"
```

psivesely replied:
It worked when I tested it.

psivesely replied:
That sugar is useful though.

conorsch commented on roles/database/tasks/initialize-raw-schema.yml:
```diff
@@ -0,0 +1,29 @@
+- name: List all schemas in the fpsd database.
+  become: true
+  become_user: postgres
+  command: psql -c '\dn' fpsd
```
This is a read-only command, meaning nothing changes on the target system, so set `changed_when: false` so it doesn't falsely report "changed". Same for two other read-only tasks listing db info in the database role.

psivesely replied:
Yup. For some reason the `postgres_user` module is not idempotent, which is why running and stroing the result of this task is necessary.

Commit:
Support PGPASSFILE in database.py

Commit:
Just unpack control_data into the crawls table

Commit:
Tweak hs_history_lookback description

Commit:
Implement util.panic(); use for critical exception handling

Commit:
Fixup get_lookback function

Commit:
Modify timestamp utility to work for db as well as logs

Commit:
Let Postgres fill empty fields for crawls table entries

Now possible because of fb6d2e39d29c9be2266fd60d5d6209693880610f.

Commit:
Minor imports and doc string cleanup

Commit:
Make all crawler methods compatible w/ database

Also, don't create folders or write files if writing to database.

Commit:
Give more general deployment dependency recs

Commit:
Support remote databases, fully idempotent db role

Commit:
Fix url_to_id_mapping dict initialization

Regression introduced by bab431f5a6e6691da60d767f425f63d5c13e51c4.

Commit:
Improve description of sorter.py, minor imports fixup

Commit:
Integrate db support into sort_onions() method

Commit:
Use get_timestamp when possible

Commit:
Publicize fetch class method; avoid limiting use of attributes

Commit:
Use dict unpacking for RawStorage.add_example() args

Commit:
Minor typos fix in crawler

Commit:
Specify full path to PGPASSFILE

Commit:
Fix crawler typos

psivesely commented on fpsd/test/test_utils.py:
```diff
@@ -0,0 +1,22 @@
+
+    def test_no_units(self):
+        with self.assertRaises(TypeError):
+            lookback_time = get_lookback('666')
```
:smiling_imp:

Commit:
PGPASSFILE env var hack

Commit:
Fix variable scope for RawStorage.get_onion_class()

Commit:
Fix handling of class tests

Commit:
Add database test

Commit:
Minor test code fixup

Commit:
Fix PGPASSFILE expr: ansible_env facts not ansible_user

Commit:
Defer string -> timedelta conversion to get_onion_class

Also, standardize naming of classes to match crawler.py

Commit:
Opt pass a db_handler obj to sorter mimicking crawler

Commit:
Fix typo in utils.get_timestamp()

Commit:
Add missing util.get_timestamp() import to database.py

Commit:
Extend get_onion_class to take variable type timespan args

Commit:
Fix sorter part of db test

Commit:
Prettier logging of class names in sorter.py

Commit:
Add utils and database tests to run_tests.py

Commit:
Only look at exact relevant timespan for db test

Commit:
Introduce test_crawl_securedrops in database test

Commit:
Create test db for use w/ testing fixtures

Commit:
Create test database for use w/ test fixtures

Create a test database with the same schemas and tables used in the real one.
This is accomplished by simply creating the test database at the end of
provisioning, and choosing the now-fully-initialized "fpsd" database as a
template.

The test database should be wiped after each use by the test fixtures, in the
tearDown() method. Hence I have introduced the wipe_database method, which, for
safety reasons, refuses to operate on database if its name starts with "test."
This method also gets tested in the database methods test during the tearDown().

Commit:
Implement utils.find_free_port()

Commit:
Add close() methods to Sorter and Crawler, explicitly kill Tor daemon

Commit:
Consolidate control_data gathering in Crawler to method

Commit:
Doc fix + name standardization

Commit:
Finish database methods test

redshiftzero commented:
Things to change: just the tests written for `get_lookback()`

Commit:
test_utils expects get_lookback's sys.exit()
